### PR TITLE
feat: support from/to paths in file patterns for extraFiles and extra…

### LINKED ISSUE #650: More advanced extraFiles
I have a project structure such as below:
```
└── MainProject
    ├── OtherProject
    │   ├── app
    │   └── dist
    ├── app
    └── dist
```
Where `OtherProject` is a separate project with its own git repository. I want to have the binary outputs of `OtherProject` to be included in the MainProject bundle, I'm using the `extraFiles` option to do so. The directory structure I want to get from my build is (for osx):
```
└── Contents
    ├── Frameworks
    ├── Info.plist
    ├── MacOS
    │   ├── mainproject
    │   └── otherproject
    └── Resources
```
But since `extraFiles` copies everything with the original directory structure relative to the project root, I can only get:
```
└── Contents
    ├── Frameworks
    ├── Info.plist
    ├── MacOS
    │   └── mainproject
    ├── OtherProject
    │   └── dist
    │       └── otherproject
    └── Resources
```

Although it's possible to get the app content structure I want by restructuring the directories before build, it's kind of an inconvenience.

I propose that aside from the glob patterns, `extraFiles` also accepts JSON objects with a structure such as:
```
{
  "from": "path/to/src",
  "to": "path/to/dst",
  "filter": ["glob patterns"]
}
```
Where `from` specifies the base directory to be copied from relative to the project root, `to` specifies the directory to be copied to relative to the app's content directory and `filter` specifies patterns to match when copying.

This would effectively solve issue #597 as well, as `from` can be a parent directory.

If you approve of this, I could send a pull request myself.

### PR DESCRIPTION
**Pull request by ha-D**

### CODE REVIEW

File: docs/Options.md [hunk 1]
```diff
@@ -210,5 +210,22 @@ May be specified in the platform options (e.g. in the `build.mac`).
 Remember that `!doNotCopyMe/**/*` would match the files *in* the `doNotCopyMe` directory, but not the directory itself, so the [empty directory](https://github.com/gulpjs/gulp/issues/165#issuecomment-32613179) would be created.
 Solution — use macro `${/*}`, e.g. `!doNotCopyMe${/*}`.
 
+## Source and Destination directories
```
Comment (develar) [inline docs/Options.md]: Please uppercase `directories` — `Directories`

File: src/platformPackager.ts [hunk 2]
```diff
@@ -306,54 +388,63 @@ export abstract class PlatformPackager<DC extends PlatformSpecificBuildOptions>
...
-  private expandPattern(pattern: string, arch: Arch): string {
-    return pattern
-      .replace(/\$\{arch}/g, Arch[arch])
-      .replace(/\$\{os}/g, this.platform.buildConfigurationKey)
-      .replace(/\$\{\/\*}/g, "{,/**/*,/**/.*}")
-  }
-
-  private doCopyExtraFiles(isResources: boolean, appOutDir: string, patterns: Array<Minimatch> | null): Promise<any> {
-    const base = isResources ? this.getResourcesDir(appOutDir) : (this.platform === Platform.MAC ? path.join(appOutDir, `${this.appInfo.productFilename}.app`, "Contents") : appOutDir)
+  private doCopyExtraFiles(isResources: boolean, patterns: Array<FileMatcher> | n): Promise<any> {
```
Comment (develar) [inline src/platformPackager.ts]: `isResources` is unused. Consider to use IntelliJ IDEA/WebStorm :)
screen shot 2016-08-11 at 07 18 33

File: src/platformPackager.ts [hunk 3]
```diff
@@ -206,24 +290,22 @@ export abstract class PlatformPackager<DC extends PlatformSpecificBuildOptions>
         rawFilter = deprecatedUserIgnoreFilter(options, this.info.appDir)
       }
 
-      const filePatterns = this.getParsedPatterns(patterns, arch)
-      let excludePatterns: Array<Minimatch> | null = null
+      let excludePatterns: FileMatcher | null = null
       if (!this.info.isTwoPackageJsonProjectLayoutUsed) {
         if (extraResourcePatterns != null) {
-          excludePatterns = extraResourcePatterns
+          excludePatterns = extraResourcePatterns[0]
```
Comment (develar) [inline src/platformPackager.ts]: This code is not correct – all patterns must be excluded.

File: src/platformPackager.ts [hunk 4]
```diff
@@ -206,24 +290,22 @@ export abstract class PlatformPackager<DC extends PlatformSpecificBuildOptions>
...
         }
         if (extraFilePatterns != null) {
           if (excludePatterns == null) {
-            excludePatterns = extraFilePatterns
+            excludePatterns = extraFilePatterns[0]
           }
           else {
-            excludePatterns = excludePatterns.concat(extraFilePatterns)
+            excludePatterns = excludePatterns.merge(extraFilePatterns[0])
```
Comment (develar) [inline src/platformPackager.ts]: This code is not correct – `cannot merge error` — see failed test.

**Comment by develar** [comment]
I added docs how to debug tests — https://github.com/electron-userland/electron-builder/blob/master/CONTRIBUTING.md#debug-tests

**Comment by develar** [comment]
Hi, do you have time ti finish PR?

**Comment by ha-D** [comment]
I do plan on finishing it, it just has to wait several days. Is that ok?

**Comment by develar** [comment]
@ha-Yes, sorry for ping.

**Comment by ha-D** [comment]
Sorry for taking long, I was preoccupied with work.
Let me know if anything still needs fixing.

**Pull request merged**